src/main/java/pl/agasior/interviewprep/controllers/TagController.java

```diff
@@ -0,0 +1,25 @@
+package pl.agasior.interviewprep.controllers;
+
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import pl.agasior.interviewprep.entities.Tag;
+import pl.agasior.interviewprep.services.tag.TagReader;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/tag")
+public class TagController {
+    private final TagReader tagReader;
+
+    TagController(TagReader tagReader) {
+        this.tagReader = tagReader;
+    }
+
+    @GetMapping
+    public ResponseEntity<List<Tag>> getTags() {
+        return ResponseEntity.ok(tagReader.getTagsByOccurrences());
```

Comment:
Same as above, both Controller class and methods doesnt need to be public

Reply:
Also, I'd suggest using TagDto (extend it with occurrencess parameter) here as response to separate http layer from persistance layer

Reply:
> Same as above, both Controller class and methods doesnt need to be public

https://sebastian.marsching.com/blog/archives/149-Springs-RequestMapping-annotation-works-on-private-methods.html

https://jira.spring.io/projects/SPR/issues/SPR-16878?filter=allissues

They can be package scoped but Sonar is suggesting for clarity of Controller to make them public.

Reply:
Not all sonar rules are worth following, take into consideration that
-Controllers are meant to be used only by http requests - Spring will handle that and noone else should be able to do that
-If you leave both class and methods package scope, you will have isolated package which clearly indicates that its not something that you want to invoke from somewhere else in the code
-Why you dont wan't to invoke it from code? A Lot of Spring functionalities doesn't work if you invoke them directly from code instead via Spring proxy (e.g @Transactional or @Valid annotations)